File: 8.cpp
// 找节点中序遍历的上一个节点
TreeNode* GetNext(TreeNode* node){
    if(node == nullptr)
        return nullptr;
    TreeNode* p = nullptr;
    if(node -> right != nullptr){
        p = node -> right;
        while(p -> left != nullptr)
            p = p -> left;
        return p;
    }
    else{
        p = node -> parent;
        if(p == nullptr)
            return nullptr;
        else if(p -> left == node){
            return p;
        }
        else if(p -> right == node){
            while(p -> parent != nullptr){
                if(p -> parent -> left == p)
                    return p -> parent;
                p = p -> parent;
            }
            return nullptr;
        }
        else
            throw std::exception();
    }
} 

File: 8.java
public class Solution {

    public TreeLinkNode GetNext(TreeLinkNode node) {

        if (node == null) {
            return null;
        }

        TreeLinkNode t = node, f = node;
        
        if (node.right != null) { // 有右子树
            t = node.right;
            while (t.left != null)
                t = t.left;
            return t;
        }
       
        // 无右子树
        f = node.next;
        while (f != null) {
            if (f.left == t) {
                return f;
            }
            if (f.right == t) {
                t = f;
                f = f.next;
            }
        }
        
        return f;
    }
}
